posix: sched: Implement sched_getparam() and sched_getscheduler()

golowanow opened the pull request:
Implement `sched_getparam()` and `sched_getscheduler()` POSIX APIs
as a part of PSE53 `_POSIX_PRIORITY_SCHEDULING` option group.

Add `CONFIG_POSIX_SCHED` Kconfig option for APIs in the group.

Add `pid_t` typedef for some configurations where it is missed and its value derived from `pthread_t`.

Resolves #66959
Resolves #66960

golowanow committed:
posix: sched: Add CONFIG_POSIX_PRIORITY_SCHEDULING

Add `CONFIG_POSIX_PRIORITY_SCHEDULING` Kconfig option to select
APIs from PSE53 `_POSIX_PRIORITY_SCHEDULING` option group.

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

cfriedt commented on include/zephyr/posix/posix_types.h:
```diff
@@ -41,6 +41,11 @@ struct pthread_attr {
 
+#if defined(CONFIG_MINIMAL_LIBC) || defined(CONFIG_ARMCLANG_STD_LIBC) \
+	|| defined(CONFIG_ARCMWDT_LIBC)
+typedef uint32_t pid_t;
```
Could this please be a signed type? Otherwise, PR looks great to me. Thank you!

golowanow replied:
done, changed to int32_t

cfriedt commented on lib/posix/sched.c:
```diff
@@ -41,3 +42,57 @@ int sched_get_priority_max(int policy)
+	 * with its first thread's id (TID).
+	 */
+	struct sched_param dummy_param = {0};
+	pthread_t tid = (pthread_t)pid;
```
Ah, I see - this is one area that is a bit touchy.

I'm not sure if it's a good idea to mix pid_t and pthread_t

golowanow replied:
yeah, it would be better to have a separate PR first to introduce `pid_t` and enable all these `pid_t` parametrized POSIX APIs, with some macro or inline functions for PID<->TID conversion based on decision for how PID value originates/relates to TID.

cfriedt replied:
@golowanow - do you feel that PID and TID are related?

cfriedt commented on lib/posix/sched.c:
```diff
@@ -41,3 +42,57 @@ int sched_get_priority_max(int policy)
+	 * with its first thread's id (TID).
+	 */
+	struct sched_param dummy_param = {0};
+	pthread_t tid = (pthread_t)pid;
```
It would be good to get some additional input here. Personally, I think that pid_t and pthread_t should not be used interchangeably.

Some current plans for pid_t can be gleaned from #65920. If you're on Discord, I would be happy to chat a bit about it

golowanow replied:
I've changed to an inline function call, which is rather a placeholder for this PR till the design decision on PID implementation.

https://github.com/zephyrproject-rtos/zephyr/pull/67028/files#diff-549102eb96e194323c2cf4a5363fe6a41aecb1978080e0117c7027f91391b298R108-R115

cfriedt replied:
Sure - well, the design decision is that Zephyr currently runs in a way that is approximately equivalent to a "single process" environment. So there would be one (maybe two at most) PIDs.

PIDs and TIDs are not considered to be equivalent. This is particularly important because processes do not share resources (by default) and have necessarily disjoint address spaces (even if not virtualized) while all threads belonging to a particular process would.

So it's possible to get the PID associated with the currently running thread by simply calling `getpid()`, but there is no simple numerical mapping between TIDs and PIDs (at this time).

That would really also necessarily be something that would need to be implemented at the kernel level, and not in the POSIX library.

cfriedt commented on lib/posix/posix_internal.h:
```diff
@@ -105,4 +105,15 @@ struct posix_thread *to_posix_thread(pthread_t pth);
+ *     must be already initialized whereas valid PID should
+ *     have positive value.
+ */
+static inline pthread_t z_get_tid(pid_t pid)
```
Personally, I would welcome proper support for separate Process IDs and processes done the right way at the kernel level, but it's probably not a good idea to try and implement a PID structure outside of the kernel.

Currently, we can only assume there is at most 1 process ID (maybe 2 in a sense) and that Zephyr runs in a single-process environment.

If you have plans to add that functionality to the kernel, it would be quite important to submit that as an RFC first.
https://github.com/zephyrproject-rtos/zephyr/issues/new/choose

golowanow replied:
> Personally, I would welcome proper support for separate Process IDs and processes done the right way at the kernel level, but it's probably not a good idea to try and implement a PID structure outside of the kernel.

I agree, that's why I'm trying to have a single wrapper function here at POSIX API level which can be later adjusted to whatever decision (or not ?) will be at the Zephyr kernel in regard of Process object and its PID going in the opposite way from POSIX standard's history perspective.

There is a bunch of POSIX APIs with `pid_t` parameters and in case of `sched_*` they need to be related to a thread, so we can assume for now that PID value maps 1:1 to the 'main' thread's id.

golowanow replied:
> > will be at the Zephyr kernel in regard of Process object and its PID going in the opposite way from POSIX standard's history perspective.
> 
> Can you please clarify?

well, I just mean that threads were accepted into POSIX after processes, as an extension, and now for Zephyr it is a kind of opposite task to implement processes having threads already.


> I get the impression that your reference in this PR is the Linux implementation of these calls.
> 
> https://linux.die.net/man/2/sched_getscheduler
> 
> Is that correct? If so, then this PR makes significantly more sense

[GNU libc](https://www.gnu.org/software/libc/manual/html_node/Process-Identification.html), Linux NTPL, and the [APUE book](https://en.wikipedia.org/wiki/Advanced_Programming_in_the_Unix_Environment) 

really, I had no intention to suggest any solution for Zephyr process design in this tiny PR )), rather a proposal to discuss it and possibly do something meaningful with these sched_get* functions instead of simply return 'not-yet-implemented'.

cfriedt commented on lib/posix/sched.c:
```diff
@@ -41,3 +43,47 @@ int sched_get_priority_max(int policy)
+ */
+int sched_getparam(pid_t pid, struct sched_param *param)
+{
+	struct sched_param dummy_param = {0};
+	pthread_t tid = 0;
+	int t_policy = -1;
+	int ret = -1;
+
+	if (param == NULL) {
+		param = &dummy_param;
+	}
+
+	tid = (pid == 0) ? pthread_self() : z_get_tid(pid);
+
+	ret = pthread_getschedparam(tid, &t_policy, param);
+	errno = ret;
+
+	return ((ret) ? -1 : 0);
```
It might be more in line with the POSIX specification to simply return `-1` and set errno to `ENOSYS` here, since Zephyr does not (yet) support processes or process scheduling.

https://pubs.opengroup.org/onlinepubs/9699919799/functions/sched_getparam.html
```suggestion
	ARG_UNUSED(pid);
	ARG_UNUSED(param);

	errno = ENOSYS;

	return -1;
```

golowanow replied:
> It might be more in line with the POSIX specification to simply return `-1` and set errno to `ENOSYS` here, since Zephyr does not (yet) support processes or process scheduling.

what do you think if we deal only with `pid==0` for `pthread_self()` and errno `ENVAL` otherwise ?

cfriedt replied:
> > It might be more in line with the POSIX specification to simply return `-1` and set errno to `ENOSYS` here, since Zephyr does not (yet) support processes or process scheduling.
> 
> what do you think if we deal only with `pid==0` for `pthread_self()` and errno `EINVAL` otherwise ?

There is still a chance for a great deal of inconsistency, even from the perspective of just querying the policy / priority.

For example, Thread A and Thread B are both threads that belong to pid zero (i.e. are part of the same process), but when one looks at them, they see that the individual thread-level scheduling policy and priority are different. So inherently, we are conflating process scheduling and thread scheduling.

The situation can even get a bit more worrisome when one considers that some threads might be running in Kernel mode, and others in user mode.

I think in order to do what Linux does (and I'm not saying it's wrong - that would be kind of nice), there would need to be support from the kernel for essentially threads being equivalent to lightweight processes (which is what Linux does), and for the kernel to support processes as well.

It's a fantastically interesting area for potential improvement, but I would not want to jump ahead of the kernel / architecture on it.

golowanow replied:
ok, removed.

> The situation can even get a bit more worrisome when one considers that some threads might be running in Kernel mode, and others in user mode.

yeah, its another design question on what to do with kernel/user threads at the posix subsys level.

cfriedt commented on lib/posix/sched.c:
```diff
@@ -41,3 +43,47 @@ int sched_get_priority_max(int policy)
+ */
+int sched_getscheduler(pid_t pid)
+{
+	struct sched_param dummy_param = {0};
+	pthread_t tid = 0;
+	int t_policy = -1;
+	int ret = -1;
+
+	tid = (pid == 0) ? pthread_self() : z_get_tid(pid);
+
+	ret = pthread_getschedparam(tid, &t_policy, &dummy_param);
+	errno = ret;
+
+	return (ret == 0) ? t_policy : -1;
```
Similarly, 

```suggestion
	ARG_UNUSED(pid);

	errno = ENOSYS;

	return -1;
```

golowanow replied:
done

cfriedt commented on tests/posix/common/src/pthread.c:
```diff
@@ -644,6 +645,87 @@ ZTEST(posix_apis, test_pthread_descriptor_leak)
+		     "failed parameter check: rc=%d, errno=%d", rc, err);
+
+	rc = sched_getscheduler(-1);
+	err = errno;
+	zassert_true((rc == -1 && err == ESRCH),
+		     "failed parameter check: rc=%d, errno=%d", rc, err);
+
+	/* .. and it is safe to call with NULL as praram. */
+	rc = sched_getparam(-1, NULL);
+	err = errno;
+	zassert_true((rc == -1 && err == ESRCH),
+		     "failed parameter check: rc=%d, errno=%d", rc, err);
+
+	/* Try with the current PID as ztest execution thread - it fails. */
+	rc = sched_getparam(0, &param);
+	err = errno;
+	zassert_true((rc == -1 && err == ESRCH),
+		     "Unexpected result : rc=%d, errno=%d", rc, err);
+
+	rc = sched_getscheduler(0);
+	err = errno;
+	zassert_true((rc == -1 && err == ESRCH),
+		     "Unexpected result : rc=%d, errno=%d", rc, err);
+
+	/* Check with a test thread. */
+	pthread_t pthread1;
+	pthread_attr_t attr1 = (pthread_attr_t){0};
+
+	param.sched_priority = sched_get_priority_min(SCHED_RR);
+	err = errno;
+	zassert_not_equal(-1, param.sched_priority,
+			  "sched_get_priority_min(SCHED_RR) failed: errno=%d", err);
+	zassert_ok(pthread_attr_init(&attr1));
+	zassert_ok(pthread_attr_setschedparam(&attr1, &param),
+			   "pthread_attr_setschedparam() failed");
+	zassert_ok(pthread_attr_setstack(&attr1, &stack_s[0][0], STACKS));
+	zassert_ok(pthread_create(&pthread1, &attr1, create_thread_sched, (void *)&param),
+			   "unable to create a test thread");
+	k_msleep(100);
+	zassert_ok(pthread_join(pthread1, NULL), "unable to join the test thread");
```
```suggestion
	zassert_equal(sched_getparam(getpid()), -1);
	zassert_equal(errno, ENOSYS);
```

golowanow replied:
done (without `getpid()` as it also not in the posix subsys yet)

cfriedt replied:
Just about there ... #67044 and #67039 need to be merged first

cfriedt reviewed (changes requested):
Quite happy for the contribution. Some changes requested to better reflect the fact that Zephyr currently does not support processes or scheduling of processes, but otherwise, looks OK.

Documentation changes should also be in a separate commit.

Thanks!

golowanow committed:
posix: sched: Implement get APIs for scheduling parameters

Initial implementation of `sched_getparam()` and `sched_getscheduler()`
POSIX APIs as a part of PSE53 `_POSIX_PRIORITY_SCHEDULING` option group.
Both functions are actually placeholders and just return `ENOSYS`
since Zephyr does not yet support processes or process scheduling.

Signed-off-by: Dmitrii Golovanov <dmitrii.golovanov@intel.com>

golowanow commented on tests/posix/common/src/pthread.c:
```diff
@@ -644,6 +645,87 @@ ZTEST(posix_apis, test_pthread_descriptor_leak)
+	}
+
+	return NULL;
+}
```
done

golowanow commented:
> Documentation changes should also be in a separate commit.

do you mean to put 'yes' here - https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/services/portability/posix/option_groups/index.rst?plain=1#L449-L450
or what is in commit with `CONFIG_POSIX_SHED` ?
isn't the former fair with `ENOSYS` functions, and the latter better to keep in the same commit ?

cfriedt commented on arch/posix/include/posix_cheats.h:
```diff
@@ -151,6 +151,8 @@ extern "C" int _posix_zephyr_main(void);
 #define sched_yield(...)		zap_sched_yield(__VA_ARGS__)
 #define sched_get_priority_min(...)	zap_sched_get_priority_min(__VA_ARGS__)
 #define sched_get_priority_max(...)	zap_sched_get_priority_max(__VA_ARGS__)
+#define sched_getparam(...)		zap_sched_getparam(__VA_ARGS__)
+#define sched_getscheduler(...)		zap_sched_getscheduler(__VA_ARGS__)
```
I don't know if this is necessary as part of this PR. I would maybe suggest making it in a separate PR.

golowanow replied:
i thought to do it same way how other APIs were introduced back then ([6 years ago!](https://github.com/zephyrproject-rtos/zephyr/pull/5667#discussion_r161603372)) keeping related changes together in the same PR 

@aescolar, what do you think ?

aescolar replied:
It is not necessary but quite harmless.
Background: This POSIX API renames were used at the time when building the POSIX API together with the host libC was allowed (the POSIX API function names were renamed with this trick to avoid them colliding with the host libC). That combination is not supported or permitted anymore, so there is no need to add more (or keep the majority of these).

golowanow replied:
> This POSIX API renames were used at the time when building the POSIX API together with the host libC was allowed (the POSIX API function names were renamed with this trick to avoid them colliding with the host libC). That combination is not supported or permitted anymore, so there is no need to add more (or keep the majority of these).

is a bulk cleanup already planned for `posix_cheats.h` ? or should it be done gradually for API groups like `sched_*` ?
in scope of this PR it looks confusing to have only some of these APIs there.

aescolar replied:
It is in my low priority TODO. So it should happen eventually, probably when I need to change something around there.

cfriedt commented on doc/services/portability/posix/aep/index.rst:
```diff
@@ -52,7 +52,7 @@ Minimal Realtime System Profile (PSE51)
     _POSIX_THREAD_PRIO_INHERIT, 200809L, :kconfig:option:`CONFIG_PTHREAD_MUTEX`
     _POSIX_THREAD_PRIO_PROTECT, -1,
-    _POSIX_THREAD_PRIORITY_SCHEDULING, -1,
+    _POSIX_THREAD_PRIORITY_SCHEDULING, -1, :kconfig:option:`CONFIG_POSIX_SCHED`
```
@golowanow, for now, since Zephyr does not support processes, we can say it's supported as far as providing a linkable symbol, but it's use falls under undefined behaviour. That's in the docs already, and it's a known issue.
```suggestion
    _POSIX_THREAD_PRIORITY_SCHEDULING, -1, :kconfig:option:`CONFIG_POSIX_SCHED` :ref:`†<posix_undefined_behaviour>`
```

golowanow replied:
done

cfriedt commented on doc/services/portability/posix/conformance/index.rst:
```diff
@@ -80,7 +80,7 @@ POSIX System Interfaces
     _POSIX_MONOTONIC_CLOCK, 200809L, :kconfig:option:`CONFIG_POSIX_CLOCK`
     _POSIX_PRIORITIZED_IO, -1,
-    :ref:`_POSIX_PRIORITY_SCHEDULING<posix_option_priority_scheduling>`, -1, :kconfig:option:`CONFIG_PTHREAD`
+    :ref:`_POSIX_PRIORITY_SCHEDULING<posix_option_priority_scheduling>`, -1, :kconfig:option:`CONFIG_POSIX_SCHED`
```
Same here, I would just 
```suggestion
    :ref:`_POSIX_PRIORITY_SCHEDULING<posix_option_priority_scheduling>`, -1, :kconfig:option:`CONFIG_POSIX_SCHED`  (will fail with ``ENOSYS``:ref:`†<posix_undefined_behaviour>`)
```

golowanow replied:
done

cfriedt commented on include/zephyr/posix/posix_types.h:
```diff
@@ -109,6 +109,12 @@ typedef struct pthread_once pthread_once_t;
 /* Newlib typedefs pthread_once_t as a struct with two ints */
 BUILD_ASSERT(sizeof(pthread_once_t) >= sizeof(struct pthread_once));
 
+/* PID */
+#if defined(CONFIG_MINIMAL_LIBC) || defined(CONFIG_ARMCLANG_STD_LIBC) \
+	|| defined(CONFIG_ARCMWDT_LIBC)
+typedef int32_t pid_t;
+#endif
+
```
Probably needs a rebase and I think this can be removed

golowanow replied:
done

cfriedt commented on lib/posix/Kconfig.sched:
```diff
@@ -0,0 +1,10 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+
+config POSIX_SCHED
```
I really ❤️  where this is going. Basically where I was headed too. One minor adjustment, if that's ok, because I'm hoping to try and make POSIX Kconfig option names map directly to POSIX Options / Groups. In the future, this might change _slightly_ though. Can chat a bit about it on Discord if you like.

```suggestion
config POSIX_PRIORITY_SCHEDULING
```

cfriedt commented on lib/posix/Kconfig.sched:
```diff
@@ -0,0 +1,10 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+
+config POSIX_SCHED
+	bool "_POSIX_PRIORITY_SCHEDULING API support"
+	default y if PTHREAD
+	depends on PTHREAD
+	help
+	  This enables POSIX scheduling APIs (_POSIX_PRIORITY_SCHEDULING).
```
I really ❤️  where this is going. Basically where I was headed too. Two minor adjustments, if that's ok, because I'm hoping to try and make POSIX Kconfig option names map directly to POSIX Options / Groups.

In the future, it might change _slightly_. Can chat a bit about it on Discord if you like.
```suggestion
config POSIX_PRIORITY_SCHEDULING
	bool "_POSIX_PRIORITY_SCHEDULING API support"
	default y if PTHREAD
	default y if POSIX_API
	depends on PTHREAD
	help
	  This enables POSIX scheduling APIs (_POSIX_PRIORITY_SCHEDULING).
```

golowanow replied:
ok, my initial intention was also to start using POSIX names here

cfriedt commented:
> > Documentation changes should also be in a separate commit.
> 
> do you mean to put 'yes' here - https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/services/portability/posix/option_groups/index.rst?plain=1#L449-L450 or what is in commit with `CONFIG_POSIX_SHED` ? is the former fair with `ENOSYS` functions, and the latter better to keep in the same commit ?

Yep - that's about it. I made some code comments to hopefully clarify.

cfriedt commented on lib/posix/Kconfig.sched:
```diff
@@ -0,0 +1,11 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+
+config POSIX_PRIORITY_SCHEDULING
```
Would it be ok if this name were more along the lines of `POSIX_SCHED`?

The main reason is because the `_POSIX_PRIORITY_SCHEDULING` is a feature test macro that should only be defined when certain features are present.

Some more info here 
https://linux.die.net/man/7/posixoptions

The goal was to tie this together with Kconfig and sysconf and there is even a PR that's getting us some of the way there, which is nice.

https://github.com/zephyrproject-rtos/zephyr/pull/67500

cfriedt replied:
Sorry, please feel free to ignore.
